### PR TITLE
fix putenv and setenv for windows

### PR DESCRIPTION
**Pull request by xmclark**
This PR separates the different platform implementations for the emscripten env functions. 

This PR attempts to fix `putenv` and `setenv`. No binding is written in libc for `putenv` so I link it. The foreign function name is `_putenv`. 

There is no such function as `setenv` in libc on windows, so we fake it by constructing a new string in the form "name=value" from the two string pointers.

There is now some duplicated code because we have the `env` module copied for windows. Look for my comments in the diff to see what actually changed.

### CODE REVIEW

File: lib/emscripten/src/env/windows/mod.rs [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+}
+
+/// emscripten: _putenv // (name: *const char);
+pub fn _putenv(name: c_int, ctx: &mut Ctx) -> c_int {
+    debug!("emscripten::_putenv");
+
+    let name_addr = emscripten_memory_pointer!(ctx.memory(0), name) as *const c_char;
+
+    debug!("=> name({:?})", unsafe { CStr::from_ptr(name_addr) });
+    unsafe { putenv(name_addr) }
```
Comment (xmclark) [inline lib/emscripten/src/env/windows/mod.rs]: this now uses the `putenv` that is declared above

File: lib/emscripten/src/env/windows/mod.rs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        let name = read_string_from_wasm(name, ctx.memory(0));
+        let value = read_string_from_wasm(value, ctx.memory(0));
+        let putenv_string = format!("{}={}", name, value);
+        let putenv_cstring = CString::from::<String>(putenv_string);
+        let putenv_raw_ptr = putenv_cstring.as_ptr();
+        debug!("=> name({:?})", CStr::from_ptr(name_addr));
+        debug!("=> value({:?})", CStr::from_ptr(value_addr));
+        putenv(putenv_raw_ptr)
+    }
+}
```
Comment (xmclark) [inline lib/emscripten/src/env/windows/mod.rs]: this was re-written and uses a helper function to pull out the strings and concatenate them

File: lib/emscripten/src/env/windows/mod.rs [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+    unsafe { copy_cstr_into_wasm(ctx, c_str) }
+}
+
+unsafe fn read_string_from_wasm(pointer: c_int, memory: &Memory) -> String {
+    let string_cell = memory.view::<u8>()[(pointer as usize)..].as_ptr();
+    let string_ptr = string_cell.as_ptr();
+    let cstr_ptr = CStr::from_ptr(string_ptr);
+    String::From(cstr_ptr)
+}
```
Comment (xmclark) [inline lib/emscripten/src/env/windows/mod.rs]: maybe this goes in utils?

File: lib/emscripten/src/env/windows/mod.rs [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+use crate::utils::{allocate_on_stack, copy_cstr_into_wasm, copy_terminated_array_of_cstrs};
+use crate::EmscriptenData;
+use wasmer_runtime_core::memory::Memory;
+use wasmer_runtime_core::vm::Ctx;
+
+#[link(name = "c")]
+extern "C" {
+    #[link_name = "_putenv"]
+    pub fn putenv(s: *const c_char) -> c_int;
+}
```
Comment (xmclark) [inline lib/emscripten/src/env/windows/mod.rs]: not actually declared in libc, so we manually declare it with an alias

File: lib/emscripten/src/env/windows/mod.rs [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+    let c_str: *mut c_char = unsafe { getenv(name_addr as _) } as _;
+    if c_str.is_null() {
+        return 0;
+    }
+
+    unsafe { copy_cstr_into_wasm(ctx, c_str) }
+}
+
+unsafe fn read_string_from_wasm(pointer: c_int, memory: &Memory) -> String {
```
Comment (lachlansneff) [inline lib/emscripten/src/env/windows/mod.rs]: I'd like to do this in a safer way actually.

```rust
fn read_string_from_wasm(memory: &Memory, pointer: u32) -> String {
    let v: Vec<_> = memory.view::<u8>()[pointer as usize]
        .iter()
        .take_while(|cell| cell.get() != 0)
        .map(|cell| cell.get())
        .collect();

    String::from_ut8(v.as_slice().unwrap()
}
```

File: lib/emscripten/src/env/windows/mod.rs (hunk 2 above)
Comment (lachlansneff) [reply lib/emscripten/src/env/windows/mod.rs]: Try to reduce the unsafe block to just the last line.

**Pull request merged**